README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Memcached components for Dart

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The Memcached module contains the following components: MemcachedLock and MemcachedCache for working with locks and cache on the Memcached server.

The module contains the following packages:
- **Build** - a standard factory for constructing components.
- **Cache** - cache Components in Memcached
- **Lock** - components of working with locks in Memcached

<a name="links"></a> Quick links:

* [Configuration](https://www.pipservices.org/recipies/configuration)
* [API Reference](https://pub.dev/documentation/pip_services3_memcached/latest/pip_services3_memcached/pip_services3_memcached-library.html)
* [Change Log](CHANGELOG.md)
* [Get Help](https://www.pipservices.org/community/help)
* [Contribute](https://www.pipservices.org/community/contribute)

## Use

Add this to your package's pubspec.yaml file:
```yaml
dependencies:
  pip_services3_memcached: version
```

Now you can install package from the command line:
```bash
pub get
```

## Develop

For development you shall install the following prerequisites:
* Dart SDK 2
* Visual Studio Code or another IDE of your choice
* Docker

Install dependencies:
```bash
pub get
```

Run automated tests:
```bash
pub run test
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized build and test as:
```bash
./build.ps1
./test.ps1
./clear.ps1
```

## Contacts

The library is created and maintained by:
- **Sergey Seroukhov**
- **Levichev Dmitry**

The documentation is written by:
- **Levichev Dmitry**

CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Memcached components for Dart Changelog

## 1.0.0

- Initial version, created by Sergey Seroukhov and Dmitry Levichev

test/fixtures/LockFixture.dart
import 'package:test/test.dart';
import 'package:pip_services3_components/pip_services3_components.dart';

final LOCK1 = 'lock_1';
final LOCK2 = 'lock_2';
final LOCK3 = 'lock_3';

class LockFixture {
  ILock _lock;

  LockFixture(ILock lock) {
    _lock = lock;
  }

  void testTryAcquireLock() async {
    // Try to acquire lock for the first time
    var result = await _lock.tryAcquireLock('123', LOCK1, 3000);
    expect(result, isTrue);

    // Try to acquire lock for the second time
    result = await _lock.tryAcquireLock('123', LOCK1, 3000);
    expect(result, isFalse);

    // Release the lock
    await _lock.releaseLock('123', LOCK1);

    // Try to acquire lock for the third time
    result = await _lock.tryAcquireLock('123', LOCK1, 3000);
    expect(result, isTrue);

    await _lock.releaseLock('123', LOCK1);
  }

  void testAcquireLock() async {
    // Acquire lock for the first time
    await _lock.acquireLock('123', LOCK2, 3000, 1000);
    // Acquire lock for the second time
    var err;
    try {
      await _lock.acquireLock('123', LOCK2, 3000, 1000);
    } catch (ex) {
      err = ex;
    }
    expect(err, isNotNull);

    // Release the lock
    await _lock.releaseLock('123', LOCK2);

    // Acquire lock for the third time
    await _lock.acquireLock('123', LOCK2, 3000, 1000);

    await _lock.releaseLock('123', LOCK2);
  }

  void testReleaseLock() async {
    // Acquire lock for the first time
    var result = await _lock.tryAcquireLock('123', LOCK3, 3000);
    expect(result, isTrue);

    // Release the lock for the first time
    await _lock.releaseLock('123', LOCK3);

    // Release the lock for the second time
    await _lock.releaseLock('123', LOCK3);
  }
}


test/fixtures/CacheFixture.dart
import 'dart:async';
import 'package:test/test.dart';
import 'package:pip_services3_components/pip_services3_components.dart';

final KEY1 = 'key1';
final KEY2 = 'key2';

final VALUE1 = 'value1';
final VALUE2 = 'value2';

class CacheFixture {
  ICache _cache;

  CacheFixture(ICache cache) {
    _cache = cache;
  }

  void testStoreAndRetrieve() async {
    await _cache.store('123', KEY1, VALUE1, 5000);

    await _cache.store('123', KEY2, VALUE2, 5000);

    await Future.delayed(Duration(milliseconds: 500));

    var val = await _cache.retrieve('123', KEY1);
    expect(val, isNotNull);
    expect(VALUE1, val);

    val = await _cache.retrieve('123', KEY2);
    expect(val, isNotNull);
    expect(VALUE2, val);
  }

  void testRetrieveExpired() async {
    await _cache.store('123', KEY1, VALUE1, 1000);

    await Future.delayed(Duration(milliseconds: 1500));

    var val = await _cache.retrieve('123', KEY1);
    expect(val, isNull);
  }

  void testRemove() async {
    await _cache.store('123', KEY1, VALUE1, 1000);

    await _cache.remove('123', KEY1);

    var val = await _cache.retrieve('123', KEY1);
    expect(val, isNull);
  }
}


test/lock/MemcachedLock.test.dart
// import { ConfigParams } from 'packages:pip_services3_commons-node';

// import { MemcachedLock } from '../../src/lock/MemcachedLock';
// import { LockFixture } from '../fixtures/LockFixture';

// suite('MemcachedLock', ()=> {
//     var _lock: MemcachedLock;
//     var _fixture: LockFixture;

//     setup((done) => {
//         let host = process.env['MEMCACHED_SERVICE_HOST'] || 'localhost';
//         let port = process.env['MEMCACHED_SERVICE_PORT'] || 11211;

//         _lock = new MemcachedLock();

//         let config = ConfigParams.fromTuples(
//             'connection.host', host,
//             'connection.port', port
//         );
//         _lock.configure(config);

//         _fixture = new LockFixture(_lock);

//         _lock.open(null, done);
//     });

//     teardown((done) => {
//         _lock.close(null, done);
//     });

//     test('Try Acquire Lock', (done) => {
//         _fixture.testTryAcquireLock(done);
//     });

//     test('Acquire Lock', (done) => {
//         _fixture.testAcquireLock(done);
//     });

//     test('Release Lock', (done) => {
//         _fixture.testReleaseLock(done);
//     });

// });


lib/src/cache/MemcachedCache.dart
import 'dart:async';

import 'package:memcache/memcache.dart';
import 'package:memcache/memcache_raw.dart';
import 'package:pip_services3_commons/pip_services3_commons.dart';
import 'package:pip_services3_components/pip_services3_components.dart';

/// Distributed cache that stores values in Memcaches caching service.
///
/// The current implementation does not support authentication.
///
/// ### Configuration parameters ###
///
/// - [connection(s)]:
///   - [discovery_key]:         (optional) a key to retrieve the connection from [IDiscovery](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/IDiscovery-class.html)
///   - [host]:                  host name or IP address
///   - [port]:                  port number
///   - [uri]:                   resource URI or connection string with all parameters in it
/// - [options]:
///   - [max_size]:              maximum number of values stored in this cache (default: 1000)
///   - [max_key_size]:          maximum key length (default: 250)
///   - [max_expiration]:        maximum expiration duration in milliseconds (default: 2592000)
///   - [max_value]:             maximum value length (default: 1048576)
///   - [pool_size]:             pool size (default: 5)
///   - [reconnect]:             reconnection timeout in milliseconds (default: 10 sec)
///   - [retries]:               number of retries (default: 3)
///   - [timeout]:               default caching timeout in milliseconds (default: 1 minute)
///   - [failures]:              number of failures before stop retrying (default: 5)
///   - [retry]:                 retry timeout in milliseconds (default: 30 sec)
///   - [idle]:                  idle timeout before disconnect in milliseconds (default: 5 sec)
///
/// ### References ###
///
/// - *:discovery:*:*:1.0        (optional) [IDiscovery](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/IDiscovery-class.html) services to resolve connection

/// ### Example ###
///
///     var cache = new MemcachedCache();
///     cache.configure(ConfigParams.fromTuples(
///       "host", "localhost",
///       "port", 11211
///     ));
///
///     cache.open("123", (err) => {
///       ...
///     });
///
///     cache.store("123", "key1", "ABC", (err) => {
///          cache.store("123", "key1", (err, value) => {
///              // Result: "ABC"
///          });
///     });

class MemcachedCache
    implements ICache, IConfigurable, IReferenceable, IOpenable {
  final _connectionResolver = ConnectionResolver();
  int _maxKeySize = 250;
  int _maxExpiration = 2592000;
  int _maxValue = 1048576;
  int _poolSize = 5;
  int _reconnect = 10000;
  int _timeout = 5000;
  int _retries = 5;
  int _failures = 5;
  int _retry = 30000;
  bool _remove = false;
  int _idle = 5000;

  Memcache _client;

  /// Creates a new instance of this cache.
  MemcachedCache();

  /// Configures component by passing configuration parameters.
  ///
  ///  -  [config]    configuration parameters to be set.
  @override
  void configure(ConfigParams config) {
    _connectionResolver.configure(config);

    _maxKeySize =
        config.getAsIntegerWithDefault('options.max_key_size', _maxKeySize);
    _maxExpiration =
        config.getAsLongWithDefault('options.max_expiration', _maxExpiration);
    _maxValue = config.getAsLongWithDefault('options.max_value', _maxValue);
    _poolSize = config.getAsIntegerWithDefault('options.pool_size', _poolSize);
    _reconnect =
        config.getAsIntegerWithDefault('options.reconnect', _reconnect);
    _timeout = config.getAsIntegerWithDefault('options.timeout', _timeout);
    _retries = config.getAsIntegerWithDefault('options.retries', _retries);
    _failures = config.getAsIntegerWithDefault('options.failures', _failures);
    _retry = config.getAsIntegerWithDefault('options.retry', _retry);
    _remove = config.getAsBooleanWithDefault('options.remove', _remove);
    _idle = config.getAsIntegerWithDefault('options.idle', _idle);
  }

  /// Sets references to dependent components.
  ///
  ///  -  [references] 	references to locate the component dependencies.
  @override
  void setReferences(IReferences references) {
    _connectionResolver.setReferences(references);
  }

  /// Checks if the component is opened.
  ///
  ///  Returns true if the component has been opened and false otherwise.
  @override
  bool isOpen() {
    return _client != null;
  }

  /// Opens the component.
  ///
  ///  -  [correlationId] 	(optional) transaction id to trace execution through call chain.
  ///  Return 			Future that receives error or null no errors occured.
  @override
  Future open(String correlationId) async {
    var connections = await _connectionResolver.resolveAll(correlationId);

    if (connections.isEmpty) {
      throw ConfigException(
          correlationId, 'NO_CONNECTION', 'Connection is not configured');
    }

    //var servers= <String>[];
    //for (var connection in connections) {
    var host = connections[0].getHost();
    var port = connections[0].getPort() ?? 11211;
    //    servers.add(host + ':' + port.toString());
    //}

    _client = Memcache.fromRaw(BinaryMemcacheProtocol(host, port));

    // var options = {
    //     maxKeySize: this._maxKeySize,
    //     maxExpiration: this._maxExpiration,
    //     maxValue: this._maxValue,
    //     poolSize: this._poolSize,
    //     reconnect: this._reconnect,
    //     timeout: this._timeout,
    //     retries: this._retries,
    //     failures: this._failures,
    //     retry: this._retry,
    //     remove: this._remove,
    //     idle: this._idle
    // };
  }

  /// Closes component and frees used resources.
  ///
  ///  -  [correlationId] 	(optional) transaction id to trace execution through call chain.
  ///  Return 			Future that receives error or null no errors occured.
  @override
  Future close(String correlationId) async {
    _client = null;
  }

  bool _checkOpened(String correlationId) {
    if (!isOpen()) {
      throw InvalidStateException(
          correlationId, 'NOT_OPENED', 'Connection is not opened');
    }
    return true;
  }

  /// Retrieves cached value from the cache using its key.
  /// If value is missing in the cache or expired it returns null.
  ///
  ///  -  [correlationId]     (optional) transaction id to trace execution through call chain.
  ///  -  [key]               a unique value key.
  ///  Return          Future that receives cached value or error.
  @override
  Future retrieve(String correlationId, String key) async {
    if (!_checkOpened(correlationId)) return;

    return await _client.get(key);
  }

  /// Stores value in the cache with expiration time.
  ///
  ///  -  [correlationId]     (optional) transaction id to trace execution through call chain.
  ///  -  [key]               a unique value key.
  ///  -  [value]             a value to store.
  ///  -  [timeout]           expiration timeout in milliseconds.
  ///  Return          (optional) Future that receives an error or null for success
  @override
  Future store(String correlationId, String key, value, int timeout) async {
    if (!_checkOpened(correlationId)) return;

    var timeoutInSec = Duration(milliseconds: timeout);
    return await _client.set(key, value, expiration: timeoutInSec);
  }

  /// Removes a value from the cache by its key.
  ///
  ///  -  [correlationId]     (optional) transaction id to trace execution through call chain.
  ///  -  [key]               a unique value key.
  ///  Return           Future that receives an error or null for success
  @override
  Future remove(String correlationId, String key) async {
    if (!_checkOpened(correlationId)) return;
    return await _client.remove(key);
  }
}


lib/src/lock/MemcachedLock.dart
//  @module lock
// import { ConfigParams } from 'packages:pip_services3_commons-node';
// import { IConfigurable } from 'packages:pip_services3_commons-node';
// import { IReferences } from 'packages:pip_services3_commons-node';
// import { IReferenceable } from 'packages:pip_services3_commons-node';
// import { IOpenable } from 'packages:pip_services3_commons-node';
// import { InvalidStateException } from 'packages:pip_services3_commons-node';
// import { ConfigException } from 'packages:pip_services3_commons-node';
// import { ConnectionResolver } from 'packages:pip_services3_components-node';
// import { Lock } from 'packages:pip_services3_components-node';

//
// /// Distributed lock that implemented based on Memcaches caching service.
// ///
// /// The current implementation does not support authentication.
// ///
// /// ### Configuration parameters ###
// ///
// /// - connection(s):
// ///   - discovery_key:         (optional) a key to retrieve the connection from [IDiscovery](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/IDiscovery-class.html)
// ///   - host:                  host name or IP address
// ///   - port:                  port number
// ///   - uri:                   resource URI or connection string with all parameters in it
// /// - options:
// ///   - retry_timeout:         timeout in milliseconds to retry lock acquisition. (Default: 100)
// ///   - max_size:              maximum number of values stored in this cache (default: 1000)
// ///   - max_key_size:          maximum key length (default: 250)
// ///   - max_expiration:        maximum expiration duration in milliseconds (default: 2592000)
// ///   - max_value:             maximum value length (default: 1048576)
// ///   - pool_size:             pool size (default: 5)
// ///   - reconnect:             reconnection timeout in milliseconds (default: 10 sec)
// ///   - retries:               number of retries (default: 3)
// ///   - timeout:               default caching timeout in milliseconds (default: 1 minute)
// ///   - failures:              number of failures before stop retrying (default: 5)
// ///   - retry:                 retry timeout in milliseconds (default: 30 sec)
// ///   - idle:                  idle timeout before disconnect in milliseconds (default: 5 sec)
// ///
// /// ### References ###
// ///
// /// - *:discovery:\*:\*:1.0        (optional) [IDiscovery](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/IDiscovery-class.html) services to resolve connection
//  *
// /// ### Example ###
// ///
// ///     let lock = new MemcachedLock();
// ///     lock.configure(ConfigParams.fromTuples(
// ///       "host", "localhost",
// ///       "port", 11211
// ///     ));
// ///
// ///     lock.open("123", (err) => {
// ///       ...
// ///     });
// ///
// ///     lock.acquire("123", "key1", (err) => {
// ///          if (err == null) {
// ///              try {
// ///                // Processing...
// ///              } finally {
// ///                 lock.releaseLock("123", "key1", (err) => {
// ///                    // Continue...
// ///                 });
// ///              }
// ///          }
// ///     });
//
// export class MemcachedLock extends Lock implements IConfigurable, IReferenceable, IOpenable {
//     private _connectionResolver: ConnectionResolver = new ConnectionResolver();

//     private _maxKeySize: number = 250;
//     private _maxExpiration: number = 2592000;
//     private _maxValue: number = 1048576;
//     private _poolSize: number = 5;
//     private _reconnect: number = 10000;
//     private _timeout: number = 5000;
//     private _retries: number = 5;
//     private _failures: number = 5;
//     private _retry: number = 30000;
//     private _remove: boolean = false;
//     private _idle: number = 5000;

//     private _client: any = null;

//
//     /// Configures component by passing configuration parameters.
//     ///
//     ///  -  config    configuration parameters to be set.
//
//     public configure(config: ConfigParams): void {
//         super.configure(config);

//         this._connectionResolver.configure(config);

//         this._maxKeySize = config.getAsIntegerWithDefault('options.max_key_size', this._maxKeySize);
//         this._maxExpiration = config.getAsLongWithDefault('options.max_expiration', this._maxExpiration);
//         this._maxValue = config.getAsLongWithDefault('options.max_value', this._maxValue);
//         this._poolSize = config.getAsIntegerWithDefault('options.pool_size', this._poolSize);
//         this._reconnect = config.getAsIntegerWithDefault('options.reconnect', this._reconnect);
//         this._timeout = config.getAsIntegerWithDefault('options.timeout', this._timeout);
//         this._retries = config.getAsIntegerWithDefault('options.retries', this._retries);
//         this._failures = config.getAsIntegerWithDefault('options.failures', this._failures);
//         this._retry = config.getAsIntegerWithDefault('options.retry', this._retry);
//         this._remove = config.getAsBooleanWithDefault('options.remove', this._remove);
//         this._idle = config.getAsIntegerWithDefault('options.idle', this._idle);
//     }

//
// 	/// Sets references to dependent components.
// 	///
// 	///  -  references 	references to locate the component dependencies.
//
//     public setReferences(references: IReferences): void {
//         this._connectionResolver.setReferences(references);
//     }

//
// 	/// Checks if the component is opened.
// 	///
// 	///  Returns true if the component has been opened and false otherwise.
//
//     public isOpen(): boolean {
//         return this._client;
//     }

//
// 	/// Opens the component.
// 	///
// 	///  -  correlationId 	(optional) transaction id to trace execution through call chain.
//     ///  -  callback 			callback function that receives error or null no errors occured.
//
//     public open(String correlationId, callback: (err: any) => void): void {
//         this._connectionResolver.resolveAll(correlationId, (err, connections) => {
//             if (err == null && connections.length == 0)
//                 err = new ConfigException(correlationId, 'NO_CONNECTION', 'Connection is not configured');

//             if (err != null) {
//                  callback(err);
//                  return;
//             }

//             let servers: string[] = [];
//             for (let connection of connections) {
//                 let host = connection.getHost();
//                 let port = connection.getPort() || 11211;
//                 servers.push(host + ':' + port);
//             }

//             let options = {
//                 maxKeySize: this._maxKeySize,
//                 maxExpiration: this._maxExpiration,
//                 maxValue: this._maxValue,
//                 poolSize: this._poolSize,
//                 reconnect: this._reconnect,
//                 timeout: this._timeout,
//                 retries: this._retries,
//                 failures: this._failures,
//                 retry: this._retry,
//                 remove: this._remove,
//                 idle: this._idle
//             };

//             let Memcached = require('memcached');
//             this._client = new Memcached(servers, options);

//             if (callback) callback(null);
//         });
//     }

//
// 	/// Closes component and frees used resources.
// 	///
// 	///  -  correlationId 	(optional) transaction id to trace execution through call chain.
//     ///  -  callback 			callback function that receives error or null no errors occured.
//
//     public close(String correlationId, callback: (err: any) => void): void {
//         this._client = null;
//         if (callback) callback(null);
//     }

//     private checkOpened(String correlationId, callback: any): boolean {
//         if (!this.isOpen()) {
//             let err = new InvalidStateException(correlationId, 'NOT_OPENED', 'Connection is not opened');
//             callback(err, null);
//             return false;
//         }

//         return true;
//     }

//
//     /// Makes a single attempt to acquire a lock by its key.
//     /// It returns immediately a positive or negative result.
//     ///
//     ///  -  correlationId     (optional) transaction id to trace execution through call chain.
//     ///  -  key               a unique lock key to acquire.
//     ///  -  ttl               a lock timeout (time to live) in milliseconds.
//     ///  -  callback          callback function that receives a lock result or error.
//
//     public tryAcquireLock(String correlationId, key: string, ttl: number,
//         callback: (err: any, result: boolean) => void): void {
//         if (!this.checkOpened(correlationId, callback)) return;

//         let lifetimeInSec = ttl / 1000;
//         this._client.add(key, 'lock', lifetimeInSec, (err) => {
//             if (err && err.message && err.message.indexOf('not stored') >= 0)
//                 callback(null, false);
//             else callback(err, err == null);
//         });
//     }

//
//     /// Releases prevously acquired lock by its key.
//     ///
//     ///  -  correlationId     (optional) transaction id to trace execution through call chain.
//     ///  -  key               a unique lock key to release.
//     ///  -  callback          callback function that receives error or null for success.
//
//     public releaseLock(String correlationId, key: string,
//         callback?: (err: any) => void): void {
//         if (!this.checkOpened(correlationId, callback)) return;

//         this._client.del(key, callback);
//     }
// }
